core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -287,6 +288,12 @@ public static RecordState forId(byte id) {
      */
     private long endOffset;
 
+    /**
+     * This maintains the set of gaps in the cachedState after initialization. Some of these gaps might correspond to batches
+     * which were fetched but not acknowledged yet. The gaps are used in determining the next fetch offset.
+     */
+    private final TreeSet<Long> gapsAfterInitialization;
```

Comment:
Is TreeSet necessary (`lg(n)` time data structure). 
You are assured that the readState reponse will always be sorted on the intervals - the order is
firstOffset, lastOffset, deliveryCount, deliveryState